examples/COCO/train_unet_coco.py

```diff
@@ -0,0 +1,193 @@
+import functools
+from typing import Tuple
+
+import flax.linen as nn
+import jax.numpy as jnp
+import time
+
+import jax
+from flax.training import train_state
+
+import numpy as np 
+import optax 
+
+
+from typing import Any, Callable, Sequence, Tuple
+import operator
+# from coco_dataset import Coco2014
+# from dataset import dataloader
+import tensorflow as tf
+from functools import partialmethod
+import utils
+import argparse
+import tqdm
+from unet import UNet
+from alpa.model.model_util import TrainState
+from alpa import (parallelize, set_parallelize_options, mark_pipeline,
+                  DeviceCluster, automatic_layer_construction)
+
+from alpa.testing import assert_allclose
+from alpa.util import get_ray_namespace_str
+import ray
+import alpa
+from alpa.util import disable_tqdm_globally
+
+# The code is to test U-net with alpa 
+# Command : python train_unet_coco.py --num-gpu 2
+
+##########################  dataset ############################
+
+# TODO(hexu): Now we are using dummy inputs, later we could enable this function for testing on real data. 
+# def get_data(batch_size):
+#     """Load COCO train and test datasets into memory."""
+#     dataset = Coco2014('/l/users/hexu.zhao/coco_dataset')
+#     train_loader, valid_loader = dataloader(dataset, batch_size=batch_size)
+#     return dataset, train_loader, valid_loader
+
+##########################  Train ##############################
+
+def train_step_jit(params, opt_state, batch):
+    """Train for a single step."""
+
+    def training_loss_fn(params, batch):
+        unet_out = model.apply(
+            params, 
+            batch[0], 
+            train=True)
+        label_one_hot = jax.nn.one_hot(batch[1], config["labels_count"])
+        loss = jnp.mean(optax.softmax_cross_entropy(logits=unet_out, labels=label_one_hot))
+        return loss
+    
+    grads = alpa.grad(training_loss_fn)(params, batch)
+    updates, opt_state = tx.update(grads, opt_state)
+    params = optax.apply_updates(params, updates)
+    return params, opt_state
+
+@parallelize(batch_argnums=(2,))
+def train_step_alpa(params, opt_state, batch):
+    """Train for a single step."""
+
+    def training_loss_fn(params, batch):
+        unet_out = model.apply(
+            params, 
+            batch[0], 
+            train=True)
+        label_one_hot = jax.nn.one_hot(batch[1], config["labels_count"])
+        loss = jnp.mean(optax.softmax_cross_entropy(logits=unet_out, labels=label_one_hot))
+        return loss
+
+    training_loss_fn = automatic_layer_construction(training_loss_fn, layer_num=4)
+    grads = alpa.grad(training_loss_fn)(params, batch)
+    updates, opt_state = tx.update(grads, opt_state)
+    params = optax.apply_updates(params, updates)
+    return params, opt_state
+
+# TODO(hexu): Now we are using dummy inputs, later we could enable this function for testing on real data. 
```

Comment:
maybe we can switch to real data when merging this PR? @merrymercy has a new dataloader which might be able to read COCO data; please check it out.

Reply:
OK, I will try using it for COCO.